File: src/main/java/kitchenpos/menus/tobe/domain/MenuProducts.java
package kitchenpos.menus.tobe.domain;

import jakarta.persistence.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

@Embeddable
public class MenuProducts {

    @OneToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinColumn(
            name = "menu_id",
            nullable = false,
            columnDefinition = "binary(16)",
            foreignKey = @ForeignKey(name = "fk_menu_product_to_menu")
    )
    private List<MenuProduct> menuProductList;

    protected MenuProducts() {
    }

    public MenuProducts(MenuProductCatalog menuProductCatalog) {
        this.menuProductList = menuProductCatalog.getEntries()
                .stream()
                .map(menuProductInfo -> new MenuProduct(menuProductInfo.id(), menuProductInfo.quantity()))
                .toList();
    }

    public List<UUID> getProductIds() {
        return menuProductList.stream()
                .map(MenuProduct::getProductId)
                .toList();
    }

    public MenuProduct getById(final UUID id) {
        return menuProductList.stream()
                .filter(menuProduct -> menuProduct.getProductId().equals(id))
                .findFirst()
                .orElseThrow(NoSuchElementException::new);
    }

    public BigDecimal calculateTotalPrice(ProductInfos productInfos) {
        return menuProductList.stream()
                .map(menuProduct -> {
                    ProductInfo productInfo = productInfos.findById(menuProduct.getProductId());
                    return productInfo.price().multiply(BigDecimal.valueOf(menuProduct.getQuantity().value()));
                })
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}

File: src/main/java/kitchenpos/menus/tobe/domain/MenuProducts.kt
package kitchenpos.menus.tobe.domain

import jakarta.persistence.CascadeType
import jakarta.persistence.Embeddable
import jakarta.persistence.FetchType
import jakarta.persistence.OneToMany
import kitchenpos.common.Price
import kitchenpos.common.ZERO
import java.util.UUID

@Embeddable
class MenuProducts(
    @OneToMany(cascade = [CascadeType.PERSIST, CascadeType.MERGE], mappedBy = "menu", fetch = FetchType.LAZY)
    val items: List<MenuProduct>
) {
    val totalPrice: Price
        get() = items.map { it.price * it.quantity.value }
            .foldRight(ZERO) { acc, price -> acc + price }

    init {
        require(items.isNotEmpty()) { "적어도 1개 이상의 상품을 등록해야합니다" }
    }

    //연관관계 편의 메소드
    fun apply(menu: Menu) {
        items.forEach { it.apply(menu) }
    }

    fun changeMenuProduct(productId: UUID, productPrice: Price) {
        val menuProduct = items.find { it.productId == productId }
            ?: throw NoSuchElementException("can not found menu product:$productId")

        menuProduct.changePrice(productPrice)
    }

    val productIds: List<UUID>
        get() = items.map { it.productId }
}
